Mathlib/NumberTheory/EulerProduct/Basic.lean

```diff
@@ -98,45 +156,107 @@ for large enough `N` gets us arbitrarily close to the sum over all natural numbe
 (assuming `f` is norm-summable and `f 0 = 0`; the latter since `0` is not smooth). -/
 lemma norm_tsum_smoothNumbers_sub_tsum_lt (hsum : Summable f) (hf₀ : f 0 = 0)
     {ε : ℝ} (εpos : 0 < ε) :
-    ∃ N₀ : ℕ, ∀ N ≥ N₀, ‖(∑' m : ℕ, f m) - (∑' m : N.smoothNumbers, f m)‖ < ε := by
-  obtain ⟨N₀, hN₀⟩ :=
-    summable_iff_nat_tsum_vanishing.mp hsum (Metric.ball 0 ε) <| Metric.ball_mem_nhds 0 εpos
-  simp_rw [mem_ball_zero_iff] at hN₀
-  refine ⟨N₀, fun N hN₁ ↦ ?_⟩
-  convert hN₀ _ <| N.smoothNumbers_compl.trans fun _ ↦ hN₁.le.trans
-  simp_rw [← tsum_subtype_add_tsum_subtype_compl hsum N.smoothNumbers,
-    add_sub_cancel_left, tsum_eq_tsum_diff_singleton (N.smoothNumbers)ᶜ hf₀]
+    ∃ N₀ : ℕ, ∀ N ≥ N₀, ‖(∑' m : ℕ, f m) - ∑' m : N.smoothNumbers, f m‖ < ε := by
+  conv => enter [1, N₀, N]; rw [smoothNumbers_eq_factoredNumbers]
+  obtain ⟨N₀, hN₀⟩ := norm_tsum_factoredNumbers_sub_tsum_lt hsum hf₀ εpos
+  refine ⟨N₀, fun N hN ↦ hN₀ (range N) fun p hp ↦ ?_⟩
+  exact mem_range.mpr <| (lt_of_mem_primesBelow hp).trans_le hN
+
 
-open Filter
+/-- The *Euler Product* for multiplicative (on coprime arguments) functions.
+
+If `f : ℕ → R`, where `R` is a complete normed commutative ring, `f 0 = 0`, `f 1 = 1`, `f` is
```

Comment:
There is something strange going on in this proof, in that `Nat.Prime` is getting unfolded to `Irreducible` at some point (which is clearly rather undesirable). Ironically, this seems to be something to do with whether or not the definition `Prime` is irreducible (as a definition)! (I swear this is not a joke.)

Reply:
There are actually *two* strange things; the other one is that the `Set` is turned into a `Prop`-valued function...
I have now added a rewrite that turns it back into something reasonable.